File: supabase/functions/_shared/shop_attribution_test.ts
import {
  assertEquals,
  assertNotEquals,
} from "https://deno.land/std@0.224.0/assert/mod.ts";
import {
  shopAttributionFromClient,
  shopAttributionFromMetadata,
  shopAttributionMetadata,
  shopAttributionQuery,
} from "./shop_attribution.ts";

const first = {
  source: "x",
  campaign: "h4_h7_pitch",
  content_id: "growth_game_t30_r1",
};

Deno.test("unavailable URL attribution is not recounted as direct", () => {
  for (const flag of [false, null, 0, "true", "false", [], {}]) {
    const payload = { ...first, attribution_valid: flag };
    assertEquals(shopAttributionFromClient(payload), null);
    assertEquals(shopAttributionMetadata(payload), null);
    assertEquals(shopAttributionQuery(payload), null);
  }
  assertEquals(
    shopAttributionFromClient({ ...first, attribution_valid: true }),
    shopAttributionFromClient(first),
  );
});

Deno.test("shop campaign and post survive the metadata round trip", () => {
  assertEquals(shopAttributionFromMetadata(shopAttributionMetadata(first)), {
    source: "x",
    campaign: "h4_h7_pitch",
    contentId: "growth_game_t30_r1",
  });
});

Deno.test("different registered posts stay distinct", () => {
  const second = { ...first, content_id: "historical_ai_t80_r1" };
  assertNotEquals(shopAttributionMetadata(first), shopAttributionMetadata(second));
  assertEquals(shopAttributionQuery(second), {
    utm_source: "x",
    utm_campaign: "h4_h7_pitch",
    utm_content: "historical_ai_t80_r1",
  });
});

Deno.test("old metadata never invents a campaign or post", () => {
  assertEquals(shopAttributionFromMetadata({ shop_source: "itch_io" }), {
    source: "itch_io",
    campaign: "",
    contentId: "",
  });
  assertEquals(shopAttributionFromClient({}), {
    source: "direct",
    campaign: "",
    contentId: "",
  });
  assertEquals(shopAttributionQuery({}), { utm_source: "direct" });
});

Deno.test("ASCII tags have the same canonical form at each boundary", () => {
  const mixed = { source: " X ", campaign: " Launch-1 ", content_id: " Post.1 " };
  assertEquals(shopAttributionFromClient(mixed), {
    source: "x",
    campaign: "launch-1",
    contentId: "post.1",
  });
  assertEquals(
    shopAttributionFromMetadata(shopAttributionMetadata(mixed)),
    shopAttributionFromClient(mixed),
  );
});

Deno.test("invalid identifiers are rejected without substitution or truncation", () => {
  for (const bad of ["a".repeat(65), "a/b", "post id", "日本語", "x@test.example"]) {
    for (const field of ["source", "campaign", "content_id"]) {
      const payload = { ...first, [field]: bad };
      assertEquals(shopAttributionFromClient(payload), null);
      assertEquals(shopAttributionMetadata(payload), null);
      assertEquals(shopAttributionQuery(payload), null);
    }
  }
  assertEquals(shopAttributionFromClient({ content_id: "a".repeat(64) })?.contentId.length, 64);
});

Deno.test("non-string tags and non-object payloads cannot become identifiers", () => {
  for (const bad of [true, 1, [], {}]) {
    assertEquals(shopAttributionFromClient({ ...first, content_id: bad }), null);
    assertEquals(shopAttributionFromMetadata({ shop_content_id: bad }), null);
  }
  for (const bad of [null, undefined, "x", 1, []]) {
    assertEquals(shopAttributionFromClient(bad), null);
    assertEquals(shopAttributionFromMetadata(bad), null);
  }
});

Deno.test("only shop tags are copied; callers retain payment and auth responsibility", () => {
  const payload = {
    ...first,
    user_id: "not-a-real-user",
    price: "1",
    stage: "purchase_complete",
    redirect: "https://untrusted.invalid/",
    email: "synthetic@example.invalid",
  };
  assertEquals(shopAttributionMetadata(payload), {
    shop_source: "x",
    shop_campaign: "h4_h7_pitch",
    shop_content_id: "growth_game_t30_r1",
  });
  assertEquals(shopAttributionQuery(payload), shopAttributionQuery(first));
  assertEquals(payload.price, "1");
});

Deno.test("unrelated metadata is not treated as shop attribution", () => {
  assertEquals(shopAttributionFromMetadata({
    utm_source: "another-source",
    utm_campaign: "another-campaign",
    utm_content: "another-post",
  }), { source: "direct", campaign: "", contentId: "" });
});

File: test/models/shop_attribution_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:my_web_app/models/shop_attribution.dart';
import 'package:my_web_app/models/shop_login_continuation.dart';

void main() {
  test('missing labels are direct with no invented campaign or post', () {
    final labels = ShopAttribution.fromUri(Uri.parse('/shop/product'));
    expect(labels.isValid, isTrue);
    expect(labels.toRequest(), {
      'source': 'direct', 'campaign': '', 'content_id': '',
    });
  });

  test('all three ASCII labels have a single canonical representation', () {
    final labels = ShopAttribution.fromUri(Uri(
      path: '/shop/product',
      queryParameters: {
        'utm_source': ' X ', 'utm_campaign': ' Launch-1 ',
        'utm_content': ' Growth_Game_T30_R1 ', 'amount': '1',
      },
    ),);
    expect(labels.toRequest(), {
      'source': 'x', 'campaign': 'launch-1',
      'content_id': 'growth_game_t30_r1',
    });
    expect(ShopAttribution.fromUri(Uri(queryParameters: labels.toQuery()))
        .toRequest(), labels.toRequest(),);
  });

  test('invalid values never alias another post or become direct', () {
    for (final value in [
      'a' * 65, 'post/a', 'post a', '日本語', 'test@example.invalid',
    ]) {
      for (final key in ['utm_source', 'utm_campaign', 'utm_content']) {
        final labels = ShopAttribution.fromUri(Uri(queryParameters: {key: value}));
        expect(labels.isValid, isFalse, reason: key);
        expect(labels.toRequest(), {'attribution_valid': false});
        expect(ShopAttribution.fromUri(Uri(queryParameters: labels.toQuery()))
            .isValid, isFalse,);
      }
    }
  });

  test('64 characters remain intact', () {
    expect(ShopAttribution.parse(contentId: 'a' * 64).contentId, 'a' * 64);
  });

  test('duplicate query keys are ambiguous even if their values match', () {
    for (final key in ['utm_source', 'utm_campaign', 'utm_content']) {
      expect(ShopAttribution.fromUri(Uri.parse('/?$key=a&$key=a')).isValid,
          isFalse,);
    }
  });

  test('server return keeps unavailable attribution unavailable', () {
    final labels = ShopAttribution.fromUri(Uri.parse(
      '/shop/product?product_id=hexciv-win64&purchase=canceled'
      '&shop_attribution=unavailable',
    ),);
    expect(labels.toRequest(), {'attribution_valid': false});
  });

  test('different posts and campaigns remain distinct', () {
    expect(ShopAttribution.parse(contentId: 'post-a').toRequest(),
        isNot(ShopAttribution.parse(contentId: 'post-b').toRequest()),);
    expect(ShopAttribution.parse(campaign: 'a').toRequest(),
        isNot(ShopAttribution.parse(campaign: 'b').toRequest()),);
  });

  test('login round trip retains only product and labels', () {
    final labels = ShopAttribution.parse(
      source: 'x', campaign: 'launch', contentId: 'post-a',
    );
    final login = ShopLoginContinuation.loginUri(
      productId: 'hexciv-win64', attribution: labels,
    );
    final product = ShopLoginContinuation.productUri(login)!;
    expect(login.path, '/login');
    expect(product.path, '/shop/product');
    expect(product.hasAuthority, isFalse);
    expect(product.queryParameters, {
      'product_id': 'hexciv-win64', ...labels.toQuery(),
    });
  });

  test('two login locations retain their own product and campaign', () {
    final first = ShopLoginContinuation.loginUri(
      productId: 'hexciv-win64',
      attribution: ShopAttribution.parse(contentId: 'post-a'),
    );
    final second = ShopLoginContinuation.loginUri(
      productId: 'another-product',
      attribution: ShopAttribution.parse(contentId: 'post-b'),
    );
    expect(ShopLoginContinuation.productUri(first)!.queryParameters,
        containsPair('utm_content', 'post-a'),);
    expect(ShopLoginContinuation.productUri(second)!.queryParameters,
        containsPair('product_id', 'another-product'),);
  });

  test('invalid labels survive login without keeping raw private values', () {
    final login = ShopLoginContinuation.loginUri(
      productId: 'hexciv-win64',
      attribution: ShopAttribution.parse(contentId: 'person@example.invalid'),
    );
    expect(login.toString(), isNot(contains('person')));
    expect(ShopAttribution.fromUri(ShopLoginContinuation.productUri(login)!)
        .isValid, isFalse,);
  });

  test('arbitrary redirect, purchase, identity and token parameters are ignored', () {
    final product = ShopLoginContinuation.productUri(Uri.parse(
      '/login?shop_product=hexciv-win64&return_to=https://evil.invalid'
      '&purchase=success&user_id=test&access_token=synthetic#secret',
    ),)!;
    expect(product.toString(),
        '/shop/product?product_id=hexciv-win64&utm_source=direct',);
  });

  test('invalid product and non-login paths cannot become a return target', () {
    for (final uri in [
      '/login', '/login?shop_product=../admin',
      '/login?shop_product=https://evil.invalid',
      '/login?shop_product=a&shop_product=b', '/other?shop_product=hexciv-win64',
    ]) {
      expect(ShopLoginContinuation.productUri(Uri.parse(uri)), isNull);
    }
    expect(ShopLoginContinuation.loginUri(
      productId: '../admin', attribution: ShopAttribution.parse(),
    ).toString(), '/login',);
  });

  test('OAuth callback uses trusted origin, not input authority or credentials', () {
    final callback = ShopLoginContinuation.callbackUri(
      base: Uri.parse('https://store.example.invalid:8443/current?token=secret'),
      loginLocation: Uri.parse(
        'https://evil.invalid/login?shop_product=hexciv-win64&utm_content=post-a'
        '&redirect=https://evil.invalid&access_token=synthetic#secret',
      ),
    )!;
    expect(callback.origin, 'https://store.example.invalid:8443');
    expect(callback.path, '/login');
    expect(callback.queryParameters, {
      'shop_product': 'hexciv-win64', 'utm_source': 'direct',
      'utm_content': 'post-a',
    });
    expect(callback.fragment, isEmpty);
    expect(callback.userInfo, isEmpty);
  });

  test('non-shop login keeps the existing default callback', () {
    expect(ShopLoginContinuation.callbackUri(
      base: Uri.parse('https://store.example.invalid'),
      loginLocation: Uri.parse('/login'),
    ), isNull,);
    expect(ShopLoginContinuation.callbackUri(
      base: Uri.parse('file:///local'),
      loginLocation: Uri.parse('/login?shop_product=hexciv-win64'),
    ), isNull,);
  });
}
